### PR TITLE
version to check

### PR DESCRIPTION
**Pull request by ksef**
Description     
The companies are all set and ready to rent out their cars. Until now, there was no opportunity to interact with potential customers in any way. Let's create a log-in option for the customers so that they can rent a car through the platform.

Objectives      
Create a table named CUSTOMER with the following columns:

ID column should be PRIMARY KEY and AUTO_INCREMENT with the type INT.
NAME column should be UNIQUE and NOT NULL with the type VARCHAR.
RENTED_CAR_ID should have the type INT. This column should be a FOREIGN KEY referring to the ID column of the CAR table, and this column can be NULL in case the customer didn't rent a car.
Update the main menu adding a couple of new options:

1. Log in as a manager
2. Log in as a customer
3. Create a customer
4. Exit

Log in as a manager should work the same as described in the previous stage.

If the option Create a customer was chosen, the program should prompt the user to enter a customer name and then save it in the database. By default, the column RENTED_CAR_ID of the customer should be NULL.

If the Log in as a customer option was chosen, the program should print the list of existing customers and prompt the user to choose one:

Choose a customer:
1. First customer
2. Second customer
3. Back
   
If the Back option was chosen, go back and print the main menu. If the customer list is empty, print the message The customer list is empty! and return to the main menu. If the user chooses one of the customers, print the customer menu:

7. Rent a car
8. Return a rented car
9. My rented car
10. Back
    
If the Back option was chosen, go back and print the main menu.

If the user chose My rented car, print the name of the car they rented and the company it belongs to:

Your rented car:        
Hyundai Venue       
Company:        
Car To Go       

In case a customer doesn't currently have any rented cars, print You didn't rent a car!.

If a customer wants to Return a rented car, the program should set the column RENTED_CAR_ID of the current customer to NULL. If the user tries to return a car that they didn't actually rent, print You didn't rent a car!.

If the customer wants to Rent a car, the program should print the list of available companies and prompt the user to choose one. If the company list is empty, the program should print the message The company list is empty! and return to the customer menu. If the customer has already rented a car, print You've already rented a car!.

If the customer chose a company that doesn't have any available cars at the moment, print No available cars in the 'Company name' company.

If the customer was lucky and the chosen company has cars to rent out, the program should print the cars ordered by their ID column and prompt the user to pick one. 

Remember that indexes in the list should start from 1:

Choose a car:
1. Hyundai Venue
2. Maruti Suzuki Dzire
3. Back

After the user has chosen a car, print the message You rented 'Car name' and then go to the customer menu. If the customer didn't like any of the options and wants to go Back, the program should take them back to the customer menu.

Note that a list numeration should always start with 1.

To pass the tests, you have to enable the auto-commit mode so that all changes are automatically saved in the database file. To do that, call the method connection.setAutoCommit(true) of the Connection object.
Examples
The greater-than symbol followed by a space > represents the user input. Note that it's not part of the input.

1. Log in as a manager
2. Log in as a customer
3. Create a customer
0. Exit
> 2

The customer list is empty!

1. Log in as a manager
2. Log in as a customer
3. Create a customer
0. Exit
> 3

Enter the customer name:
> First customer
The customer was added!

1. Log in as a manager
2. Log in as a customer
3. Create a customer
0. Exit
> 3

Enter the customer name:
> Second customer
The customer was added!

1. Log in as a manager
2. Log in as a customer
3. Create a customer
0. Exit
> 2

Customer list:
1. First customer
2. Second customer
0. Back
> 1

1. Rent a car
2. Return a rented car
3. My rented car
0. Back
> 3

You didn't rent a car!

1. Rent a car
2. Return a rented car
3. My rented car
0. Back
> 2

You didn't rent a car!

1. Rent a car
2. Return a rented car
3. My rented car
0. Back
> 0

1. Log in as a manager
2. Log in as a customer
3. Create a customer
0. Exit
> 0

### CODE REVIEW

File: build.gradle [hunk 1]
```diff
@@ -5,15 +5,32 @@ plugins {
...
+}
+
 repositories {
     mavenCentral()
 }
 
 dependencies {
-    testImplementation 'org.junit.jupiter:junit-jupiter-api:5.7.2'
-    testRuntimeOnly 'org.junit.jupiter:junit-jupiter-engine:5.7.2'
+    testImplementation 'org.junit.jupiter:junit-jupiter-api:5.8.2'
```
Comment (taras-nishchymenko) [inline build.gradle]: Please group your implementation and test specific dependencies

File: src/main/java/carsharing/CarSharingEngine.java [hunk 2]
```diff
@@ -0,0 +1,90 @@
...
+import carsharing.customer.CustomerDAOImpl;
+import carsharing.roles.CustomerRole;
+import carsharing.roles.ManagerRole;
+import carsharing.configuration.DBManager;
+
+import java.util.Scanner;
+
+public class CarSharingEngine {
+
+    public static final String BACK = "0. Back";
```
Comment (taras-nishchymenko) [inline src/main/java/carsharing/CarSharingEngine.java]: Looks weird. Do we need it?

File: src/main/java/carsharing/CarSharingEngine.java [hunk 3]
```diff
@@ -0,0 +1,90 @@
...
+    private final Scanner scanner;
+    private final CustomerDAO customerDao;
+    private String pathToDb;
```
Comment (taras-nishchymenko) [inline src/main/java/carsharing/CarSharingEngine.java]: what is this?

File: src/main/java/carsharing/CarSharingEngine.java [hunk 4]
```diff
@@ -0,0 +1,90 @@
...
+    private final DBManager DBManager;
```
Comment (taras-nishchymenko) [inline src/main/java/carsharing/CarSharingEngine.java]: why do we need it?

File: src/main/java/carsharing/configuration/DBManager.java [hunk 5]
```diff
@@ -0,0 +1,60 @@
...
+
+    private static final String CREATE_TABLE_CUSTOMER = """
+            CREATE TABLE IF NOT EXISTS customer
+            (id INTEGER PRIMARY KEY AUTO_INCREMENT, name VARCHAR(100) UNIQUE NOT NULL,
+            rented_car_id INTEGER DEFAULT NULL,
+            CONSTRAINT fk_car FOREIGN KEY(rented_car_id)
+            REFERENCES car(id));
+            """;
+
+    protected static String DB_URL;
```
Comment (taras-nishchymenko) [inline src/main/java/carsharing/configuration/DBManager.java]: Why do we need to change state of static field?

File: src/main/java/carsharing/configuration/DBManager.java [hunk 6]
```diff
@@ -0,0 +1,60 @@
...
+
+import lombok.NoArgsConstructor;
+
+import java.sql.Connection;
+import java.sql.DriverManager;
+import java.sql.SQLException;
+import java.sql.Statement;
+
+@NoArgsConstructor
+public class DBManager {
```
Comment (taras-nishchymenko) [inline src/main/java/carsharing/configuration/DBManager.java:11]: Why our DAOs extend this class?

File: src/main/java/carsharing/customer/CustomerDAOImpl.java [hunk 7]
```diff
@@ -0,0 +1,214 @@
...
+    private static final String UPDATE_CUSTOMER_WITH_NULL_RENTED_CAR = "UPDATE customer SET rented_car_id = NULL WHERE id = ?";
+    private static final String SELECT_ID_BY_CAR_NAME = "SELECT id FROM car WHERE name = ?";
+    private static final String UPDATE_CUSTOMER = "UPDATE customer SET rented_car_id = ? WHERE id = ?";
+    private static final String SELECT_ID_BY_COMPANY = "SELECT id FROM company WHERE name = ?";
+    private static final String SELECT_RENTED_CAR = "SELECT rented_car_id FROM customer WHERE rented_car_id IS NOT NULL";
+    private static final String SELECT_ID_NAME_BY_CAR = "SELECT id, name FROM car WHERE company_id = ? ORDER BY id";
+    private static final String INSERT_INTO_CUSTOMER_BY_NAME = "INSERT INTO customer (name) VALUES ?";
+
+    private final DBManager DBManager;
+    private Set<Integer> rentedCars;
```
Comment (taras-nishchymenko) [inline src/main/java/carsharing/customer/CustomerDAOImpl.java]: Why do we need store state outside of DB?

File: src/main/java/carsharing/customer/CustomerDAOImpl.java [hunk 8]
```diff
@@ -0,0 +1,214 @@
...
+    private static final String SELECT_NAME_BY_COMPANY = "SELECT name FROM company WHERE id = ?";
+    private static final String UPDATE_CUSTOMER_WITH_NULL_RENTED_CAR = "UPDATE customer SET rented_car_id = NULL WHERE id = ?";
+    private static final String SELECT_ID_BY_CAR_NAME = "SELECT id FROM car WHERE name = ?";
+    private static final String UPDATE_CUSTOMER = "UPDATE customer SET rented_car_id = ? WHERE id = ?";
+    private static final String SELECT_ID_BY_COMPANY = "SELECT id FROM company WHERE name = ?";
+    private static final String SELECT_RENTED_CAR = "SELECT rented_car_id FROM customer WHERE rented_car_id IS NOT NULL";
+    private static final String SELECT_ID_NAME_BY_CAR = "SELECT id, name FROM car WHERE company_id = ? ORDER BY id";
+    private static final String INSERT_INTO_CUSTOMER_BY_NAME = "INSERT INTO customer (name) VALUES ?";
+
+    private final DBManager DBManager;
```
Comment (taras-nishchymenko) [inline src/main/java/carsharing/customer/CustomerDAOImpl.java]: Why do we need DBManager if this class extends DBManager?

File: src/main/java/carsharing/customer/CustomerDAOImpl.java [hunk 9]
```diff
@@ -0,0 +1,214 @@
...
+                    setCompany(customerId, resultSet, customer);
+                }
+            }
+        } catch (SQLException e) {
+            e.printStackTrace();
+        }
+        return customer;
+    }
+
+    private void setCompanyId(Customer customer, int carId) {
```
Comment (taras-nishchymenko) [inline src/main/java/carsharing/customer/CustomerDAOImpl.java]: please place all private methods after public

File: src/main/java/carsharing/customer/CustomerDAOImpl.java [hunk 10]
```diff
@@ -0,0 +1,214 @@
...
+        try (Connection connection = DBManager.getConnection();
+             PreparedStatement preparedStatement = connection.prepareStatement(SELECT_NAME_AND_CAR_ID)) {
+            connection.setAutoCommit(true);
+            preparedStatement.setInt(1, customerId);
+            ResultSet resultSet = preparedStatement.executeQuery();
+            while (resultSet.next()) {
+                customer = new Customer(resultSet.getString(1));
+                int carId = resultSet.getInt(2);
+                if (carId != 0) {
+                    setCompanyId(customer, carId);
```
Comment (taras-nishchymenko) [inline src/main/java/carsharing/customer/CustomerDAOImpl.java]: Please avoid such shadow operations. It’s better to find company and set it here

File: src/main/java/carsharing/customer/CustomerDAOImpl.java [hunk 11]
```diff
@@ -0,0 +1,214 @@
...
+                }
+            }
+        } catch (SQLException e) {
+            e.printStackTrace();
+        }
+        return customer;
+    }
+
+    private void setCompanyId(Customer customer, int carId) {
+        try (Connection connection = DBManager.getConnection();
```
Comment (taras-nishchymenko) [inline src/main/java/carsharing/customer/CustomerDAOImpl.java]: We already have opened connection when using this method

File: src/main/java/carsharing/customer/CustomerDAOImpl.java (hunk 10 above)
Comment (taras-nishchymenko) [reply src/main/java/carsharing/customer/CustomerDAOImpl.java]: And consider joins instead of using several select queries

**Review by taras-nishchymenko — CHANGES_REQUESTED** [review]

File: src/main/java/carsharing/CarSharingEngine.java (hunk 2 above)
Comment (ksef) [reply src/main/java/carsharing/CarSharingEngine.java]: deleted

File: src/main/java/carsharing/CarSharingEngine.java (hunk 3 above)
Comment (ksef) [reply src/main/java/carsharing/CarSharingEngine.java]: this is db name, renamed

File: src/main/java/carsharing/CarSharingEngine.java (hunk 4 above)
Comment (ksef) [reply src/main/java/carsharing/CarSharingEngine.java]: It moved this to constructor. 
I need it to pass the dbName to DBManager

File: src/main/java/carsharing/configuration/DBManager.java (hunk 5 above)
Comment (ksef) [reply src/main/java/carsharing/configuration/DBManager.java]: DB_URL = "jdbc:h2:./src/main/java/carsharing/db/" + name; 
We can change db name

File: src/main/java/carsharing/configuration/DBManager.java (hunk 6 above)
Comment (ksef) [reply src/main/java/carsharing/configuration/DBManager.java:11]: I fixed this

File: src/main/java/carsharing/customer/CustomerDAOImpl.java (hunk 7 above)
Comment (ksef) [reply src/main/java/carsharing/customer/CustomerDAOImpl.java]: This is car id, we need it to assign a car id.
I fix this

File: src/main/java/carsharing/customer/CustomerDAOImpl.java (hunk 8 above)
Comment (ksef) [reply src/main/java/carsharing/customer/CustomerDAOImpl.java]: in new version all DAO classes doesn't extends DBManager

File: src/main/java/carsharing/customer/CustomerDAOImpl.java (hunk 9 above)
Comment (ksef) [reply src/main/java/carsharing/customer/CustomerDAOImpl.java]: placed

File: src/main/java/carsharing/customer/CustomerDAOImpl.java (hunk 11 above)
Comment (ksef) [reply src/main/java/carsharing/customer/CustomerDAOImpl.java]: fixed

File: build.gradle (hunk 1 above)
Comment (ksef) [reply build.gradle]: grouped

**Pull request closed without merging**